File: scala-finch/src/main/scala/org/openapitools/apis/DefaultApi.scala
package org.openapitools.apis

import java.io._
import traveltimeplatform._
import org.openapitools.models._
import org.openapitools.models.RequestRoutes
import org.openapitools.models.RequestSupportedLocations
import org.openapitools.models.RequestTimeFilter
import org.openapitools.models.RequestTimeFilterFast
import org.openapitools.models.RequestTimeFilterPostcodeDistricts
import org.openapitools.models.RequestTimeFilterPostcodeSectors
import org.openapitools.models.RequestTimeFilterPostcodes
import org.openapitools.models.RequestTimeMap
import org.openapitools.models.ResponseError
import org.openapitools.models.ResponseGeocoding
import org.openapitools.models.ResponseMapInfo
import org.openapitools.models.ResponseRoutes
import org.openapitools.models.ResponseSupportedLocations
import org.openapitools.models.ResponseTimeFilter
import org.openapitools.models.ResponseTimeFilterFast
import org.openapitools.models.ResponseTimeFilterPostcodeDistricts
import org.openapitools.models.ResponseTimeFilterPostcodeSectors
import org.openapitools.models.ResponseTimeFilterPostcodes
import org.openapitools.models.ResponseTimeMap
import io.finch.circe._
import io.circe.generic.semiauto._
import com.twitter.concurrent.AsyncStream
import com.twitter.finagle.Service
import com.twitter.finagle.Http
import com.twitter.finagle.http.{Request, Response}
import com.twitter.finagle.http.exp.Multipart.{FileUpload, InMemoryFileUpload, OnDiskFileUpload}
import com.twitter.util.Future
import com.twitter.io.Buf
import io.finch._, items._
import java.io.File
import java.time._

object DefaultApi {
    /**
    * Compiles all service endpoints.
    * @return Bundled compilation of all service endpoints.
    */
    def endpoints(da: DataAccessor) =
        geocodingReverseSearch(da) :+:
        geocodingSearch(da) :+:
        mapInfo(da) :+:
        routes(da) :+:
        supportedLocations(da) :+:
        timeFilter(da) :+:
        timeFilterFast(da) :+:
        timeFilterPostcodeDistricts(da) :+:
        timeFilterPostcodeSectors(da) :+:
        timeFilterPostcodes(da) :+:
        timeMap(da)


    private def checkError(e: CommonError) = e match {
      case InvalidInput(_) => BadRequest(e)
      case MissingIdentifier(_) => BadRequest(e)
      case RecordNotFound(_) => NotFound(e)
      case _ => InternalServerError(e)
    }

    implicit class StringOps(s: String) {

      import java.time.format.DateTimeFormatter

      lazy val localformatter: DateTimeFormatter = DateTimeFormatter.ofPattern("yyyy-MM-dd")
      lazy val datetimeformatter: DateTimeFormatter =
      DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss.SSSZ")

      def toLocalDateTime: LocalDateTime = LocalDateTime.parse(s,localformatter)
      def toZonedDateTime: ZonedDateTime = ZonedDateTime.parse(s, datetimeformatter)

    }

        /**
        * 
        * @return An endpoint representing a ResponseGeocoding
        */
        private def geocodingReverseSearch(da: DataAccessor): Endpoint[ResponseGeocoding] =
        get("v4" :: "geocoding" :: "reverse" :: param("focus.lat").map(_.toDouble) :: param("focus.lng").map(_.toDouble) :: paramOption("within.country") :: header("X-Api-Key") :: header("X-Application-Id")) { (focusPeriodlat: Double, focusPeriodlng: Double, withinPeriodcountry: Option[String], authParamApiKey: String, authParamApplicationId: String) =>
          da.Default_geocodingReverseSearch(focusPeriodlat, focusPeriodlng, withinPeriodcountry, authParamApiKey, authParamApplicationId) match {
            case Left(error) => checkError(error)
            case Right(data) => Ok(data)
          }
        } handle {
          case e: Exception => BadRequest(e)
        }

        /**
        * 
        * @return An endpoint representing a ResponseGeocoding
        */
        private def geocodingSearch(da: DataAccessor): Endpoint[ResponseGeocoding] =
        get("v4" :: "geocoding" :: "search" :: param("query") :: paramOption("within.country") :: paramOption("focus.lat").map(_.map(_.toDouble)) :: paramOption("focus.lng").map(_.map(_.toDouble)) :: header("X-Api-Key") :: header("X-Application-Id")) { (query: String, withinPeriodcountry: Option[String], focusPeriodlat: Option[Double], focusPeriodlng: Option[Double], authParamApiKey: String, authParamApplicationId: String) =>
          da.Default_geocodingSearch(query, withinPeriodcountry, focusPeriodlat, focusPeriodlng, authParamApiKey, authParamApplicationId) match {
            case Left(error) => checkError(error)
            case Right(data) => Ok(data)
          }
        } handle {
          case e: Exception => BadRequest(e)
        }

        /**
        * 
        * @return An endpoint representing a ResponseMapInfo
        */
        private def mapInfo(da: DataAccessor): Endpoint[ResponseMapInfo] =
        get("v4" :: "map-info" :: header("X-Api-Key") :: header("X-Application-Id")) { (authParamApiKey: String, authParamApplicationId: String) =>
          da.Default_mapInfo(authParamApiKey, authParamApplicationId) match {
            case Left(error) => checkError(error)
            case Right(data) => Ok(data)
          }
        } handle {
          case e: Exception => BadRequest(e)
        }

        /**
        * 
        * @return An endpoint representing a ResponseRoutes
        */
        private def routes(da: DataAccessor): Endpoint[ResponseRoutes] =
        post("v4" :: "routes" :: jsonBody[RequestRoutes] :: header("X-Api-Key") :: header("X-Application-Id")) { (requestRoutes: RequestRoutes, authParamApiKey: String, authParamApplicationId: String) =>
          da.Default_routes(requestRoutes, authParamApiKey, authParamApplicationId) match {
            case Left(error) => checkError(error)
            case Right(data) => Ok(data)
          }
        } handle {
          case e: Exception => BadRequest(e)
        }

        /**
        * 
        * @return An endpoint representing a ResponseSupportedLocations
        */
        private def supportedLocations(da: DataAccessor): Endpoint[ResponseSupportedLocations] =
        post("v4" :: "supported-locations" :: jsonBody[RequestSupportedLocations] :: header("X-Api-Key") :: header("X-Application-Id")) { (requestSupportedLocations: RequestSupportedLocations, authParamApiKey: String, authParamApplicationId: String) =>
          da.Default_supportedLocations(requestSupportedLocations, authParamApiKey, authParamApplicationId) match {
            case Left(error) => checkError(error)
            case Right(data) => Ok(data)
          }
        } handle {
          case e: Exception => BadRequest(e)
        }

        /**
        * 
        * @return An endpoint representing a ResponseTimeFilter
        */
        private def timeFilter(da: DataAccessor): Endpoint[ResponseTimeFilter] =
        post("v4" :: "time-filter" :: jsonBody[RequestTimeFilter] :: header("X-Api-Key") :: header("X-Application-Id")) { (requestTimeFilter: RequestTimeFilter, authParamApiKey: String, authParamApplicationId: String) =>
          da.Default_timeFilter(requestTimeFilter, authParamApiKey, authParamApplicationId) match {
            case Left(error) => checkError(error)
            case Right(data) => Ok(data)
          }
        } handle {
          case e: Exception => BadRequest(e)
        }

        /**
        * 
        * @return An endpoint representing a ResponseTimeFilterFast
        */
        private def timeFilterFast(da: DataAccessor): Endpoint[ResponseTimeFilterFast] =
        post("v4" :: "time-filter" :: "fast" :: jsonBody[RequestTimeFilterFast] :: header("X-Api-Key") :: header("X-Application-Id")) { (requestTimeFilterFast: RequestTimeFilterFast, authParamApiKey: String, authParamApplicationId: String) =>
          da.Default_timeFilterFast(requestTimeFilterFast, authParamApiKey, authParamApplicationId) match {
            case Left(error) => checkError(error)
            case Right(data) => Ok(data)
          }
        } handle {
          case e: Exception => BadRequest(e)
        }

        /**
        * 
        * @return An endpoint representing a ResponseTimeFilterPostcodeDistricts
        */
        private def timeFilterPostcodeDistricts(da: DataAccessor): Endpoint[ResponseTimeFilterPostcodeDistricts] =
        post("v4" :: "time-filter" :: "postcode-districts" :: jsonBody[RequestTimeFilterPostcodeDistricts] :: header("X-Api-Key") :: header("X-Application-Id")) { (requestTimeFilterPostcodeDistricts: RequestTimeFilterPostcodeDistricts, authParamApiKey: String, authParamApplicationId: String) =>
          da.Default_timeFilterPostcodeDistricts(requestTimeFilterPostcodeDistricts, authParamApiKey, authParamApplicationId) match {
            case Left(error) => checkError(error)
            case Right(data) => Ok(data)
          }
        } handle {
          case e: Exception => BadRequest(e)
        }

        /**
        * 
        * @return An endpoint representing a ResponseTimeFilterPostcodeSectors
        */
        private def timeFilterPostcodeSectors(da: DataAccessor): Endpoint[ResponseTimeFilterPostcodeSectors] =
        post("v4" :: "time-filter" :: "postcode-sectors" :: jsonBody[RequestTimeFilterPostcodeSectors] :: header("X-Api-Key") :: header("X-Application-Id")) { (requestTimeFilterPostcodeSectors: RequestTimeFilterPostcodeSectors, authParamApiKey: String, authParamApplicationId: String) =>
          da.Default_timeFilterPostcodeSectors(requestTimeFilterPostcodeSectors, authParamApiKey, authParamApplicationId) match {
            case Left(error) => checkError(error)
            case Right(data) => Ok(data)
          }
        } handle {
          case e: Exception => BadRequest(e)
        }

        /**
        * 
        * @return An endpoint representing a ResponseTimeFilterPostcodes
        */
        private def timeFilterPostcodes(da: DataAccessor): Endpoint[ResponseTimeFilterPostcodes] =
        post("v4" :: "time-filter" :: "postcodes" :: jsonBody[RequestTimeFilterPostcodes] :: header("X-Api-Key") :: header("X-Application-Id")) { (requestTimeFilterPostcodes: RequestTimeFilterPostcodes, authParamApiKey: String, authParamApplicationId: String) =>
          da.Default_timeFilterPostcodes(requestTimeFilterPostcodes, authParamApiKey, authParamApplicationId) match {
            case Left(error) => checkError(error)
            case Right(data) => Ok(data)
          }
        } handle {
          case e: Exception => BadRequest(e)
        }

        /**
        * 
        * @return An endpoint representing a ResponseTimeMap
        */
        private def timeMap(da: DataAccessor): Endpoint[ResponseTimeMap] =
        post("v4" :: "time-map" :: jsonBody[RequestTimeMap] :: header("X-Api-Key") :: header("X-Application-Id")) { (requestTimeMap: RequestTimeMap, authParamApiKey: String, authParamApplicationId: String) =>
          da.Default_timeMap(requestTimeMap, authParamApiKey, authParamApplicationId) match {
            case Left(error) => checkError(error)
            case Right(data) => Ok(data)
          }
        } handle {
          case e: Exception => BadRequest(e)
        }


    implicit private def fileUploadToFile(fileUpload: FileUpload) : File = {
      fileUpload match {
        case upload: InMemoryFileUpload =>
          bytesToFile(Buf.ByteArray.Owned.extract(upload.content))
        case upload: OnDiskFileUpload =>
          upload.content
        case _ => null
      }
    }

    private def bytesToFile(input: Array[Byte]): java.io.File = {
      val file = File.createTempFile("tmpDefaultApi", null)
      val output = new FileOutputStream(file)
      output.write(input)
      file
    }

    // This assists in params(string) application (which must be Seq[A] in parameter list) when the param is used as a List[A] elsewhere.
    implicit def seqList[A](input: Seq[A]): List[A] = input.toList
}

File: csharp-dotnet2/src/main/CsharpDotNet2/traveltimeplatform/Api/DefaultApi.cs
using System;
using System.Collections.Generic;
using RestSharp;
using Org.OpenAPITools.Client;
using traveltimeplatform.Model;

namespace traveltimeplatform.Api
{
    /// <summary>
    /// Represents a collection of functions to interact with the API endpoints
    /// </summary>
    public interface IDefaultApi
    {
        /// <summary>
        ///  
        /// </summary>
        /// <param name="focusLat"></param>
        /// <param name="focusLng"></param>
        /// <param name="withinCountry"></param>
        /// <returns>ResponseGeocoding</returns>
        ResponseGeocoding GeocodingReverseSearch (double? focusLat, double? focusLng, string withinCountry);
        /// <summary>
        ///  
        /// </summary>
        /// <param name="query"></param>
        /// <param name="withinCountry"></param>
        /// <param name="focusLat"></param>
        /// <param name="focusLng"></param>
        /// <returns>ResponseGeocoding</returns>
        ResponseGeocoding GeocodingSearch (string query, string withinCountry, double? focusLat, double? focusLng);
        /// <summary>
        ///  
        /// </summary>
        /// <returns>ResponseMapInfo</returns>
        ResponseMapInfo MapInfo ();
        /// <summary>
        ///  
        /// </summary>
        /// <param name="requestRoutes"></param>
        /// <returns>ResponseRoutes</returns>
        ResponseRoutes Routes (RequestRoutes requestRoutes);
        /// <summary>
        ///  
        /// </summary>
        /// <param name="requestSupportedLocations"></param>
        /// <returns>ResponseSupportedLocations</returns>
        ResponseSupportedLocations SupportedLocations (RequestSupportedLocations requestSupportedLocations);
        /// <summary>
        ///  
        /// </summary>
        /// <param name="requestTimeFilter"></param>
        /// <returns>ResponseTimeFilter</returns>
        ResponseTimeFilter TimeFilter (RequestTimeFilter requestTimeFilter);
        /// <summary>
        ///  
        /// </summary>
        /// <param name="requestTimeFilterFast"></param>
        /// <returns>ResponseTimeFilterFast</returns>
        ResponseTimeFilterFast TimeFilterFast (RequestTimeFilterFast requestTimeFilterFast);
        /// <summary>
        ///  
        /// </summary>
        /// <param name="requestTimeFilterPostcodeDistricts"></param>
        /// <returns>ResponseTimeFilterPostcodeDistricts</returns>
        ResponseTimeFilterPostcodeDistricts TimeFilterPostcodeDistricts (RequestTimeFilterPostcodeDistricts requestTimeFilterPostcodeDistricts);
        /// <summary>
        ///  
        /// </summary>
        /// <param name="requestTimeFilterPostcodeSectors"></param>
        /// <returns>ResponseTimeFilterPostcodeSectors</returns>
        ResponseTimeFilterPostcodeSectors TimeFilterPostcodeSectors (RequestTimeFilterPostcodeSectors requestTimeFilterPostcodeSectors);
        /// <summary>
        ///  
        /// </summary>
        /// <param name="requestTimeFilterPostcodes"></param>
        /// <returns>ResponseTimeFilterPostcodes</returns>
        ResponseTimeFilterPostcodes TimeFilterPostcodes (RequestTimeFilterPostcodes requestTimeFilterPostcodes);
        /// <summary>
        ///  
        /// </summary>
        /// <param name="requestTimeMap"></param>
        /// <returns>ResponseTimeMap</returns>
        ResponseTimeMap TimeMap (RequestTimeMap requestTimeMap);
    }
  
    /// <summary>
    /// Represents a collection of functions to interact with the API endpoints
    /// </summary>
    public class DefaultApi : IDefaultApi
    {
        /// <summary>
        /// Initializes a new instance of the <see cref="DefaultApi"/> class.
        /// </summary>
        /// <param name="apiClient"> an instance of ApiClient (optional)</param>
        /// <returns></returns>
        public DefaultApi(ApiClient apiClient = null)
        {
            if (apiClient == null) // use the default one in Configuration
                this.ApiClient = Configuration.DefaultApiClient; 
            else
                this.ApiClient = apiClient;
        }
    
        /// <summary>
        /// Initializes a new instance of the <see cref="DefaultApi"/> class.
        /// </summary>
        /// <returns></returns>
        public DefaultApi(String basePath)
        {
            this.ApiClient = new ApiClient(basePath);
        }
    
        /// <summary>
        /// Sets the base path of the API client.
        /// </summary>
        /// <param name="basePath">The base path</param>
        /// <value>The base path</value>
        public void SetBasePath(String basePath)
        {
            this.ApiClient.BasePath = basePath;
        }
    
        /// <summary>
        /// Gets the base path of the API client.
        /// </summary>
        /// <param name="basePath">The base path</param>
        /// <value>The base path</value>
        public String GetBasePath(String basePath)
        {
            return this.ApiClient.BasePath;
        }
    
        /// <summary>
        /// Gets or sets the API client.
        /// </summary>
        /// <value>An instance of the ApiClient</value>
        public ApiClient ApiClient {get; set;}
    
        /// <summary>
        ///  
        /// </summary>
        /// <param name="focusLat"></param> 
        /// <param name="focusLng"></param> 
        /// <param name="withinCountry"></param> 
        /// <returns>ResponseGeocoding</returns>            
        public ResponseGeocoding GeocodingReverseSearch (double? focusLat, double? focusLng, string withinCountry)
        {
            
            // verify the required parameter 'focusLat' is set
            if (focusLat == null) throw new ApiException(400, "Missing required parameter 'focusLat' when calling GeocodingReverseSearch");
            
            // verify the required parameter 'focusLng' is set
            if (focusLng == null) throw new ApiException(400, "Missing required parameter 'focusLng' when calling GeocodingReverseSearch");
            
    
            var path = "/v4/geocoding/reverse";
            path = path.Replace("{format}", "json");
                
            var queryParams = new Dictionary<String, String>();
            var headerParams = new Dictionary<String, String>();
            var formParams = new Dictionary<String, String>();
            var fileParams = new Dictionary<String, FileParameter>();
            String postBody = null;
    
             if (focusLat != null) queryParams.Add("focus.lat", ApiClient.ParameterToString(focusLat)); // query parameter
 if (focusLng != null) queryParams.Add("focus.lng", ApiClient.ParameterToString(focusLng)); // query parameter
 if (withinCountry != null) queryParams.Add("within.country", ApiClient.ParameterToString(withinCountry)); // query parameter
                                        
            // authentication setting, if any
            String[] authSettings = new String[] { "ApiKey", "ApplicationId" };
    
            // make the HTTP request
            IRestResponse response = (IRestResponse) ApiClient.CallApi(path, Method.GET, queryParams, postBody, headerParams, formParams, fileParams, authSettings);
    
            if (((int)response.StatusCode) >= 400)
                throw new ApiException ((int)response.StatusCode, "Error calling GeocodingReverseSearch: " + response.Content, response.Content);
            else if (((int)response.StatusCode) == 0)
                throw new ApiException ((int)response.StatusCode, "Error calling GeocodingReverseSearch: " + response.ErrorMessage, response.ErrorMessage);
    
            return (ResponseGeocoding) ApiClient.Deserialize(response.Content, typeof(ResponseGeocoding), response.Headers);
        }
    
        /// <summary>
        ///  
        /// </summary>
        /// <param name="query"></param> 
        /// <param name="withinCountry"></param> 
        /// <param name="focusLat"></param> 
        /// <param name="focusLng"></param> 
        /// <returns>ResponseGeocoding</returns>            
        public ResponseGeocoding GeocodingSearch (string query, string withinCountry, double? focusLat, double? focusLng)
        {
            
            // verify the required parameter 'query' is set
            if (query == null) throw new ApiException(400, "Missing required parameter 'query' when calling GeocodingSearch");
            
    
            var path = "/v4/geocoding/search";
            path = path.Replace("{format}", "json");
                
            var queryParams = new Dictionary<String, String>();
            var headerParams = new Dictionary<String, String>();
            var formParams = new Dictionary<String, String>();
            var fileParams = new Dictionary<String, FileParameter>();
            String postBody = null;
    
             if (query != null) queryParams.Add("query", ApiClient.ParameterToString(query)); // query parameter
 if (withinCountry != null) queryParams.Add("within.country", ApiClient.ParameterToString(withinCountry)); // query parameter
 if (focusLat != null) queryParams.Add("focus.lat", ApiClient.ParameterToString(focusLat)); // query parameter
 if (focusLng != null) queryParams.Add("focus.lng", ApiClient.ParameterToString(focusLng)); // query parameter
                                        
            // authentication setting, if any
            String[] authSettings = new String[] { "ApiKey", "ApplicationId" };
    
            // make the HTTP request
            IRestResponse response = (IRestResponse) ApiClient.CallApi(path, Method.GET, queryParams, postBody, headerParams, formParams, fileParams, authSettings);
    
            if (((int)response.StatusCode) >= 400)
                throw new ApiException ((int)response.StatusCode, "Error calling GeocodingSearch: " + response.Content, response.Content);
            else if (((int)response.StatusCode) == 0)
                throw new ApiException ((int)response.StatusCode, "Error calling GeocodingSearch: " + response.ErrorMessage, response.ErrorMessage);
    
            return (ResponseGeocoding) ApiClient.Deserialize(response.Content, typeof(ResponseGeocoding), response.Headers);
        }
    
        /// <summary>
        ///  
        /// </summary>
        /// <returns>ResponseMapInfo</returns>            
        public ResponseMapInfo MapInfo ()
        {
            
    
            var path = "/v4/map-info";
            path = path.Replace("{format}", "json");
                
            var queryParams = new Dictionary<String, String>();
            var headerParams = new Dictionary<String, String>();
            var formParams = new Dictionary<String, String>();
            var fileParams = new Dictionary<String, FileParameter>();
            String postBody = null;
    
                                                    
            // authentication setting, if any
            String[] authSettings = new String[] { "ApiKey", "ApplicationId" };
    
            // make the HTTP request
            IRestResponse response = (IRestResponse) ApiClient.CallApi(path, Method.GET, queryParams, postBody, headerParams, formParams, fileParams, authSettings);
    
            if (((int)response.StatusCode) >= 400)
                throw new ApiException ((int)response.StatusCode, "Error calling MapInfo: " + response.Content, response.Content);
            else if (((int)response.StatusCode) == 0)
                throw new ApiException ((int)response.StatusCode, "Error calling MapInfo: " + response.ErrorMessage, response.ErrorMessage);
    
            return (ResponseMapInfo) ApiClient.Deserialize(response.Content, typeof(ResponseMapInfo), response.Headers);
        }
    
        /// <summary>
        ///  
        /// </summary>
        /// <param name="requestRoutes"></param> 
        /// <returns>ResponseRoutes</returns>            
        public ResponseRoutes Routes (RequestRoutes requestRoutes)
        {
            
            // verify the required parameter 'requestRoutes' is set
            if (requestRoutes == null) throw new ApiException(400, "Missing required parameter 'requestRoutes' when calling Routes");
            
    
            var path = "/v4/routes";
            path = path.Replace("{format}", "json");
                
            var queryParams = new Dictionary<String, String>();
            var headerParams = new Dictionary<String, String>();
            var formParams = new Dictionary<String, String>();
            var fileParams = new Dictionary<String, FileParameter>();
            String postBody = null;
    
                                                postBody = ApiClient.Serialize(requestRoutes); // http body (model) parameter
    
            // authentication setting, if any
            String[] authSettings = new String[] { "ApiKey", "ApplicationId" };
    
            // make the HTTP request
            IRestResponse response = (IRestResponse) ApiClient.CallApi(path, Method.POST, queryParams, postBody, headerParams, formParams, fileParams, authSettings);
    
            if (((int)response.StatusCode) >= 400)
                throw new ApiException ((int)response.StatusCode, "Error calling Routes: " + response.Content, response.Content);
            else if (((int)response.StatusCode) == 0)
                throw new ApiException ((int)response.StatusCode, "Error calling Routes: " + response.ErrorMessage, response.ErrorMessage);
    
            return (ResponseRoutes) ApiClient.Deserialize(response.Content, typeof(ResponseRoutes), response.Headers);
        }
    
        /// <summary>
        ///  
        /// </summary>
        /// <param name="requestSupportedLocations"></param> 
        /// <returns>ResponseSupportedLocations</returns>            
        public ResponseSupportedLocations SupportedLocations (RequestSupportedLocations requestSupportedLocations)
        {
            
            // verify the required parameter 'requestSupportedLocations' is set
            if (requestSupportedLocations == null) throw new ApiException(400, "Missing required parameter 'requestSupportedLocations' when calling SupportedLocations");
            
    
            var path = "/v4/supported-locations";
            path = path.Replace("{format}", "json");
                
            var queryParams = new Dictionary<String, String>();
            var headerParams = new Dictionary<String, String>();
            var formParams = new Dictionary<String, String>();
            var fileParams = new Dictionary<String, FileParameter>();
            String postBody = null;
    
                                                postBody = ApiClient.Serialize(requestSupportedLocations); // http body (model) parameter
    
            // authentication setting, if any
            String[] authSettings = new String[] { "ApiKey", "ApplicationId" };
    
            // make the HTTP request
            IRestResponse response = (IRestResponse) ApiClient.CallApi(path, Method.POST, queryParams, postBody, headerParams, formParams, fileParams, authSettings);
    
            if (((int)response.StatusCode) >= 400)
                throw new ApiException ((int)response.StatusCode, "Error calling SupportedLocations: " + response.Content, response.Content);
            else if (((int)response.StatusCode) == 0)
                throw new ApiException ((int)response.StatusCode, "Error calling SupportedLocations: " + response.ErrorMessage, response.ErrorMessage);
    
            return (ResponseSupportedLocations) ApiClient.Deserialize(response.Content, typeof(ResponseSupportedLocations), response.Headers);
        }
    
        /// <summary>
        ///  
        /// </summary>
        /// <param name="requestTimeFilter"></param> 
        /// <returns>ResponseTimeFilter</returns>            
        public ResponseTimeFilter TimeFilter (RequestTimeFilter requestTimeFilter)
        {
            
            // verify the required parameter 'requestTimeFilter' is set
            if (requestTimeFilter == null) throw new ApiException(400, "Missing required parameter 'requestTimeFilter' when calling TimeFilter");
            
    
            var path = "/v4/time-filter";
            path = path.Replace("{format}", "json");
                
            var queryParams = new Dictionary<String, String>();
            var headerParams = new Dictionary<String, String>();
            var formParams = new Dictionary<String, String>();
            var fileParams = new Dictionary<String, FileParameter>();
            String postBody = null;
    
                                                postBody = ApiClient.Serialize(requestTimeFilter); // http body (model) parameter
    
            // authentication setting, if any
            String[] authSettings = new String[] { "ApiKey", "ApplicationId" };
    
            // make the HTTP request
            IRestResponse response = (IRestResponse) ApiClient.CallApi(path, Method.POST, queryParams, postBody, headerParams, formParams, fileParams, authSettings);
    
            if (((int)response.StatusCode) >= 400)
                throw new ApiException ((int)response.StatusCode, "Error calling TimeFilter: " + response.Content, response.Content);
            else if (((int)response.StatusCode) == 0)
                throw new ApiException ((int)response.StatusCode, "Error calling TimeFilter: " + response.ErrorMessage, response.ErrorMessage);
    
            return (ResponseTimeFilter) ApiClient.Deserialize(response.Content, typeof(ResponseTimeFilter), response.Headers);
        }
    
        /// <summary>
        ///  
        /// </summary>
        /// <param name="requestTimeFilterFast"></param> 
        /// <returns>ResponseTimeFilterFast</returns>            
        public ResponseTimeFilterFast TimeFilterFast (RequestTimeFilterFast requestTimeFilterFast)
        {
            
            // verify the required parameter 'requestTimeFilterFast' is set
            if (requestTimeFilterFast == null) throw new ApiException(400, "Missing required parameter 'requestTimeFilterFast' when calling TimeFilterFast");
            
    
            var path = "/v4/time-filter/fast";
            path = path.Replace("{format}", "json");
                
            var queryParams = new Dictionary<String, String>();
            var headerParams = new Dictionary<String, String>();
            var formParams = new Dictionary<String, String>();
            var fileParams = new Dictionary<String, FileParameter>();
            String postBody = null;
    
                                                postBody = ApiClient.Serialize(requestTimeFilterFast); // http body (model) parameter
    
            // authentication setting, if any
            String[] authSettings = new String[] { "ApiKey", "ApplicationId" };
    
            // make the HTTP request
            IRestResponse response = (IRestResponse) ApiClient.CallApi(path, Method.POST, queryParams, postBody, headerParams, formParams, fileParams, authSettings);
    
            if (((int)response.StatusCode) >= 400)
                throw new ApiException ((int)response.StatusCode, "Error calling TimeFilterFast: " + response.Content, response.Content);
            else if (((int)response.StatusCode) == 0)
                throw new ApiException ((int)response.StatusCode, "Error calling TimeFilterFast: " + response.ErrorMessage, response.ErrorMessage);
    
            return (ResponseTimeFilterFast) ApiClient.Deserialize(response.Content, typeof(ResponseTimeFilterFast), response.Headers);
        }
    
        /// <summary>
        ///  
        /// </summary>
        /// <param name="requestTimeFilterPostcodeDistricts"></param> 
        /// <returns>ResponseTimeFilterPostcodeDistricts</returns>            
        public ResponseTimeFilterPostcodeDistricts TimeFilterPostcodeDistricts (RequestTimeFilterPostcodeDistricts requestTimeFilterPostcodeDistricts)
        {
            
            // verify the required parameter 'requestTimeFilterPostcodeDistricts' is set
            if (requestTimeFilterPostcodeDistricts == null) throw new ApiException(400, "Missing required parameter 'requestTimeFilterPostcodeDistricts' when calling TimeFilterPostcodeDistricts");
            
    
            var path = "/v4/time-filter/postcode-districts";
            path = path.Replace("{format}", "json");
                
            var queryParams = new Dictionary<String, String>();
            var headerParams = new Dictionary<String, String>();
            var formParams = new Dictionary<String, String>();
            var fileParams = new Dictionary<String, FileParameter>();
            String postBody = null;
    
                                                postBody = ApiClient.Serialize(requestTimeFilterPostcodeDistricts); // http body (model) parameter
    
            // authentication setting, if any
            String[] authSettings = new String[] { "ApiKey", "ApplicationId" };
    
            // make the HTTP request
            IRestResponse response = (IRestResponse) ApiClient.CallApi(path, Method.POST, queryParams, postBody, headerParams, formParams, fileParams, authSettings);
    
            if (((int)response.StatusCode) >= 400)
                throw new ApiException ((int)response.StatusCode, "Error calling TimeFilterPostcodeDistricts: " + response.Content, response.Content);
            else if (((int)response.StatusCode) == 0)
                throw new ApiException ((int)response.StatusCode, "Error calling TimeFilterPostcodeDistricts: " + response.ErrorMessage, response.ErrorMessage);
    
            return (ResponseTimeFilterPostcodeDistricts) ApiClient.Deserialize(response.Content, typeof(ResponseTimeFilterPostcodeDistricts), response.Headers);
        }
    
        /// <summary>
        ///  
        /// </summary>
        /// <param name="requestTimeFilterPostcodeSectors"></param> 
        /// <returns>ResponseTimeFilterPostcodeSectors</returns>            
        public ResponseTimeFilterPostcodeSectors TimeFilterPostcodeSectors (RequestTimeFilterPostcodeSectors requestTimeFilterPostcodeSectors)
        {
            
            // verify the required parameter 'requestTimeFilterPostcodeSectors' is set
            if (requestTimeFilterPostcodeSectors == null) throw new ApiException(400, "Missing required parameter 'requestTimeFilterPostcodeSectors' when calling TimeFilterPostcodeSectors");
            
    
            var path = "/v4/time-filter/postcode-sectors";
            path = path.Replace("{format}", "json");
                
            var queryParams = new Dictionary<String, String>();
            var headerParams = new Dictionary<String, String>();
            var formParams = new Dictionary<String, String>();
            var fileParams = new Dictionary<String, FileParameter>();
            String postBody = null;
    
                                                postBody = ApiClient.Serialize(requestTimeFilterPostcodeSectors); // http body (model) parameter
    
            // authentication setting, if any
            String[] authSettings = new String[] { "ApiKey", "ApplicationId" };
    
            // make the HTTP request
            IRestResponse response = (IRestResponse) ApiClient.CallApi(path, Method.POST, queryParams, postBody, headerParams, formParams, fileParams, authSettings);
    
            if (((int)response.StatusCode) >= 400)
                throw new ApiException ((int)response.StatusCode, "Error calling TimeFilterPostcodeSectors: " + response.Content, response.Content);
            else if (((int)response.StatusCode) == 0)
                throw new ApiException ((int)response.StatusCode, "Error calling TimeFilterPostcodeSectors: " + response.ErrorMessage, response.ErrorMessage);
    
            return (ResponseTimeFilterPostcodeSectors) ApiClient.Deserialize(response.Content, typeof(ResponseTimeFilterPostcodeSectors), response.Headers);
        }
    
        /// <summary>
        ///  
        /// </summary>
        /// <param name="requestTimeFilterPostcodes"></param> 
        /// <returns>ResponseTimeFilterPostcodes</returns>            
        public ResponseTimeFilterPostcodes TimeFilterPostcodes (RequestTimeFilterPostcodes requestTimeFilterPostcodes)
        {
            
            // verify the required parameter 'requestTimeFilterPostcodes' is set
            if (requestTimeFilterPostcodes == null) throw new ApiException(400, "Missing required parameter 'requestTimeFilterPostcodes' when calling TimeFilterPostcodes");
            
    
            var path = "/v4/time-filter/postcodes";
            path = path.Replace("{format}", "json");
                
            var queryParams = new Dictionary<String, String>();
            var headerParams = new Dictionary<String, String>();
            var formParams = new Dictionary<String, String>();
            var fileParams = new Dictionary<String, FileParameter>();
            String postBody = null;
    
                                                postBody = ApiClient.Serialize(requestTimeFilterPostcodes); // http body (model) parameter
    
            // authentication setting, if any
            String[] authSettings = new String[] { "ApiKey", "ApplicationId" };
    
            // make the HTTP request
            IRestResponse response = (IRestResponse) ApiClient.CallApi(path, Method.POST, queryParams, postBody, headerParams, formParams, fileParams, authSettings);
    
            if (((int)response.StatusCode) >= 400)
                throw new ApiException ((int)response.StatusCode, "Error calling TimeFilterPostcodes: " + response.Content, response.Content);
            else if (((int)response.StatusCode) == 0)
                throw new ApiException ((int)response.StatusCode, "Error calling TimeFilterPostcodes: " + response.ErrorMessage, response.ErrorMessage);
    
            return (ResponseTimeFilterPostcodes) ApiClient.Deserialize(response.Content, typeof(ResponseTimeFilterPostcodes), response.Headers);
        }
    
        /// <summary>
        ///  
        /// </summary>
        /// <param name="requestTimeMap"></param> 
        /// <returns>ResponseTimeMap</returns>            
        public ResponseTimeMap TimeMap (RequestTimeMap requestTimeMap)
        {
            
            // verify the required parameter 'requestTimeMap' is set
            if (requestTimeMap == null) throw new ApiException(400, "Missing required parameter 'requestTimeMap' when calling TimeMap");
            
    
            var path = "/v4/time-map";
            path = path.Replace("{format}", "json");
                
            var queryParams = new Dictionary<String, String>();
            var headerParams = new Dictionary<String, String>();
            var formParams = new Dictionary<String, String>();
            var fileParams = new Dictionary<String, FileParameter>();
            String postBody = null;
    
                                                postBody = ApiClient.Serialize(requestTimeMap); // http body (model) parameter
    
            // authentication setting, if any
            String[] authSettings = new String[] { "ApiKey", "ApplicationId" };
    
            // make the HTTP request
            IRestResponse response = (IRestResponse) ApiClient.CallApi(path, Method.POST, queryParams, postBody, headerParams, formParams, fileParams, authSettings);
    
            if (((int)response.StatusCode) >= 400)
                throw new ApiException ((int)response.StatusCode, "Error calling TimeMap: " + response.Content, response.Content);
            else if (((int)response.StatusCode) == 0)
                throw new ApiException ((int)response.StatusCode, "Error calling TimeMap: " + response.ErrorMessage, response.ErrorMessage);
    
            return (ResponseTimeMap) ApiClient.Deserialize(response.Content, typeof(ResponseTimeMap), response.Headers);
        }
    
    }
}
